xptifw/include/xpti_int64_hash_table.hpp

```diff
@@ -0,0 +1,280 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+#pragma once
+#include "xpti_data_types.h"
+
+#include <atomic>
+#include <mutex>
+#include <unordered_map>
+
+#ifdef XPTI_STATISTICS
+#include <stdio.h>
+#endif
+
+#ifdef XPTI_USE_TBB
+#include <tbb/concurrent_hash_map.h>
+#include <tbb/spin_mutex.h>
+
+namespace xpti {
+/// \brief A class for mapping one 64-bit value to another 64-bit value
+/// \details With each payload, a kernel/function name and the source file name
+/// may be passed and we need to ensure that the payload can be cached in a hash
+/// map that maps a unique value from the payload to a universal ID. We could
+/// use the payload hash for this purpose, but the numbers are non-monotonic and
+/// can be harder to debug. This implementation of the hash table uses Threading
+/// Building Blocks concurrent containers for multi-threaded efficiency.
+class Hash64x64Table {
+public:
+  using ht_lut_t = tbb::concurrent_hash_map<int64_t, int64_t>;
+
+  Hash64x64Table(int size = 1024)
+      : MForward(size), MReverse(size), MTableSize(size) {
+#ifdef XPTI_STATISTICS
+    MInsertions = 0;
+    MRetrievals = 0;
+#endif
+  }
+
+  ~Hash64x64Table() {
+    MForward.clear();
+    MReverse.clear();
+  }
+
+  //  Clear all the contents of this hash table and get it ready for re-use
```

Comment:
Should we convert all such comments to doxygen?

Reply:
None of the APIs here in the hash table should be visible through the XPTI interfaces. If there's a strong need, I can address it later.